File: prozess_synchronisation/loesung/02/schlafender_friseur_lockfile.py
import threading
import time
import os

FILENAME = "friseur.lock"
waiting_room_counter = 0

barber_sleep = False

def create_lockfile():
    while True:
        if not os.path.exists(FILENAME):
            with open(FILENAME, "w") as outfile:
                pass
            break 
        # else: es wird gewartet (Ausführung blockiert), bis Lockfile frei ist

def remove_lockfile():
    os.remove(FILENAME)

def barber():
    global waiting_room_counter
    global barber_sleep
    time.sleep(0.2) # initiale Verzögerung (öffnet den Laden)
    while True:
        if not barber_sleep:
            if waiting_room_counter > 0:
                waiting_room_counter -= 1
                print("[Friseur] schneidet Haare ...")
            create_lockfile()
            if waiting_room_counter == 0:
                print("[Friseur] geht schlafen ...")
                time.sleep(1) # Interrupt emulieren
                barber_sleep = True 
            remove_lockfile()
    

def customer(counter):
    global waiting_room_counter
    global barber_sleep
    print(f"\t[Kunde #{counter}] betritt den Saloon.")
    waiting_room_counter += 1
    if waiting_room_counter == 1:
        print(f"\t[Kunde #{counter}] weckt Friseur auf.")
        create_lockfile()
        barber_sleep = False
        remove_lockfile()

# Threads starten
threading.Thread(target=barber).start()
for i in range(10):
    threading.Thread(target=customer, args=(i+1,)).start()
    time.sleep(0.5)

File: prozess_synchronisation/loesung/02/schlafender_friseur_lockfile.cpp
#include <iostream>
#include <thread>
#include <chrono>
#include <fstream>
#include <string>
#include <unistd.h>  // für access()
#include <cstdio>    // für remove()

using namespace std;

const string FILENAME = "friseur.lock";
int waiting_room_counter = 0;
bool barber_sleep = false;

void create_lockfile() {
    while (true) {
        // prüfen, ob Lockfile existiert
        if (access(FILENAME.c_str(), F_OK) != 0) {
            ofstream outfile(FILENAME);
            outfile.close();
            break;
        }
        // sonst: warten (Ausführung blockiert), bis Lockfile frei ist
    }
}

void remove_lockfile() {
    remove(FILENAME.c_str());
}

void barber() {
    this_thread::sleep_for(chrono::milliseconds(200));
    while (true) {
        if (!barber_sleep) {
            if (waiting_room_counter > 0) {
                waiting_room_counter--;
                cout << "[Friseur] schneidet Haare ..." << endl;
            }
            create_lockfile();
            if (waiting_room_counter == 0) {
                cout << "[Friseur] geht schlafen ..." << endl;
                this_thread::sleep_for(chrono::seconds(1)); // Interrupt emulieren
                barber_sleep = true;
            }
            remove_lockfile();
        }
    }
}

void customer(int counter) {
    cout << "\t[Kunde #" << counter << "] betritt den Saloon." << endl;
    waiting_room_counter++;
    if (waiting_room_counter == 1) {
        cout << "\t[Kunde #" << counter << "] weckt Friseur auf." << endl;
        create_lockfile();
        barber_sleep = false;
        remove_lockfile();
    }
}

int main() {
    thread barber_thread(barber);

    for (int i = 0; i < 10; i++) {
        thread customer_thread(customer, i+1);
        customer_thread.join();
        this_thread::sleep_for(chrono::milliseconds(500));
    }

    barber_thread.detach();

    return 0;
}
